### PR TITLE
Update user info

### PR DESCRIPTION
**Pull request by ProDog**

### CODE REVIEW

**Review by Lichen9618 — APPROVED** [review]

File: contracts/IdoContract/src/IdoContract.cs [hunk 1]
```diff
@@ -321,20 +282,74 @@ public static ByteString[] GetAllRegisteredProjects()
...
+            return allContracts.Count;
+        }
+
+        private static void SetProjectUserInfo(UInt160 idoPairContractHash, UInt160 user, ProjectUserInfo projectUserInfo)
+        {
+            byte[] key = projectUserPrefix.Concat(idoPairContractHash).Concat(user);
+            Storage.Put(Storage.CurrentContext, key, StdLib.Serialize(projectUserInfo));
+        }
+
+        public static ProjectUserInfo GetProjectUserInfo(UInt160 idoPairContractHash, UInt160 user)
```
Comment (shargon) [inline contracts/IdoContract/src/IdoContract.cs:305]: The argument was not checked, it must be checked to be Valid and !=Zero

File: contracts/IdoContract/src/IdoContract.cs [hunk 2]
```diff
@@ -411,79 +422,27 @@ public static bool SwapTokenSecondRound(UInt160 user, UInt160 idoPairContractHas
...
-        }
+            ProjectUserInfo projectUserInfo = GetProjectUserInfo(idoPairContractHash, user);
+            ExecutionEngine.Assert(projectUserInfo.claimAmount > 0, "no unclaimed token");
 
-        public static BigInteger GetRegisteredProjectUserWeight(UInt160 idoPairContractHash, UInt160 user)
-        {
-            byte[] userVoteKey = userVotePrefix.Concat(idoPairContractHash).Concat(user);
-            ByteString rawVoted = Storage.Get(Storage.CurrentContext, userVoteKey);
-            return rawVoted is null ? 0 : (BigInteger)rawVoted;
+            SafeTransfer(project.tokenHash, Runtime.ExecutingScriptHash, user, projectUserInfo.claimAmount);
```
Comment (shargon) [inline contracts/IdoContract/src/IdoContract.cs:435]: Do the transfer at the end, after set the flags

File: contracts/IdoContract/src/IdoContract.cs [hunk 3]
```diff
@@ -382,9 +384,18 @@ public static bool SwapToken(UInt160 user, UInt160 idoPairContractHash, BigInteg
             SafeTransfer(GetSpendAssetHash(), user, Runtime.ExecutingScriptHash, spendAssetAmount);
             SafeTransfer(GetSpendAssetHash(), Runtime.ExecutingScriptHash, idoPairContractHash, spendAssetAmount);
             BigInteger balanceAfter = GetBalanceOfToken(project.tokenHash, Runtime.ExecutingScriptHash);
-            //ExecutionEngine.Assert(balanceAfter - balanceBefore == amount, "amount not correct");
-            AddUserClaimAmount(idoPairContractHash, user, amount);
-            AddUserSwapAmount(idoPairContractHash, user, amount);
+
+            BigInteger recAmount = balanceAfter - balanceBefore;
+            ExecutionEngine.Assert(recAmount > 0, "amount not correct");
```
Comment (shargon) [inline contracts/IdoContract/src/IdoContract.cs:389]: `>=` ?

Comment (ProDog) [reply contracts/IdoContract/src/IdoContract.cs:389]: swapped amount == 0?

Comment (ProDog) [reply contracts/IdoContract/src/IdoContract.cs:389]: I think the amount of swap can't be == 0

File: contracts/IdoContract/src/IdoContract.cs (hunk 2 above)
Comment (ProDog) [reply contracts/IdoContract/src/IdoContract.cs:435]: OK, I will fix it in new pr.

**Pull request merged**